### PR TITLE
Relational Database Shredder

### LINKED ISSUE #3031: RDB Shredder: rename from Scala Hadoop Shred

### PR DESCRIPTION
**Pull request by BenFradet**
Fixes #3031

### CODE REVIEW

**Comment by alexanderdean** [comment]
* Commit title should be `Relational Database Shredder: rename from Scala Hadoop Shred (close #3037)`
* Missing an update to this: https://github.com/snowplow/snowplow/blob/master/3-enrich/scala-hadoop-shred/project/SnowplowHadoopShredBuild.scala#L27

**Comment by BenFradet** [comment]
Updated :+1:

**Comment by alexanderdean** [comment]
Looks good - you can widen this branch to include some of the other tickets and add commits into this same branch; no need to silo each ticket into a separate PR...

**Comment by alexanderdean** [comment]
Assigning this PR into R9x [HAD] Scala Hadoop Shred reboot

**Comment by BenFradet** [comment]
Oh, ok, welp I'll pick commits between branches.

**Comment by BenFradet** [comment]
for the spark job:
- [ ] finish the job
- [ ] doc
- [ ] test
- [ ] rm the scalding job

File: 4-storage/relational-database-shredder/src/main/scala/com.snowplowanalytics.snowplow.storage/hadoop/utils/JsonUtils.scala [hunk 1]
```diff
@@ -12,7 +12,7 @@
  */
 package com.snowplowanalytics
 package snowplow
-package enrich
+package storage
 package hadoop
```
Comment (alexanderdean) [inline 4-storage/relational-database-shredder/src/main/scala/com.snowplowanalytics.snowplow.storage/hadoop/utils/JsonUtils.scala]: package spark or hadoop?

Comment (BenFradet) [inline 4-storage/relational-database-shredder/src/main/scala/com.snowplowanalytics.snowplow.storage/hadoop/utils/JsonUtils.scala]: I left the old code in place, this is an artifact from the change of package #3036

**Comment by alexanderdean** [comment]
It's coming along nicely! Looks like you still have the big refactoring out of the PartitionTap to do.

I saw a mix of `hadoop` and `spark` for the package names, maybe that update isn't finished yet...

**Comment by BenFradet** [comment]
As we discussed with @chuwy , so far I have an output dir which looks something like:

```
shredded/
    good/
        run=2016-11-26-21-48-42/
            atomic-events/
                part-00000
                part-00001
                ...
            vendor=com.amce/
                name=event/
                    format=jsonschema/
                        version=1-0-0/
                            part-00001-00010
                name=context/
                    format=jsonschema/
                        version=1-0-0/
                            part-00001-00010
```

instead of

```
shredded/
    good/
        run=2016-11-26-21-48-42/
            atomic-events/
                part-00000
                part-00001
                ...
            com.amce/
                event/
                    jsonschema/
                        1-0-0/
                            part-00001-00010
                context/
                    jsonschema/
                        1-0-0/
                            part-00001-00010
```

@alexanderdean would that be a problem?

**Comment by alexanderdean** [comment]
Interesting - I think we've actually had users ask for a structure more like this (because it's more Hiveable). I don't see this as a dealbreaker - we can upgrade StorageLoader in tandem to handle the new structure...

**Comment by BenFradet** [comment]
Found another issue compared to the existing when writing the tests, in Spark you can't write to a directory that has already been written to. This is the case when writing shredded events to the parent of the atomic events directory.

One way to circumvent this issue would to have the following directory structure:

```
shredded/
    good/
        run=2016-11-26-21-48-42/
            atomic-events/
                part-00000
                part-00001
                ...
            shredded-events/
                vendor=com.amce/
                    name=event/
                        format=jsonschema/
                            version=1-0-0/
                                part-00001-00010
                    name=context/
                        format=jsonschema/
                            version=1-0-0/
                                part-00001-00010
```

Notice the added `shredded_events` layer.

**Comment by alexanderdean** [comment]
That sounds fine to me:

1. Remember to make sister tickets for the attendant StorageLoader changes in the same milestone
2. I would call it `shredded-types` rather than `shredded-events` as it contains contexts/entities as well as events

**Comment by BenFradet** [comment]
created #3044

**Comment by BenFradet** [comment]
Since we're upgrading sbt, we might as well convert to a .sbt-based build instead of .scala, what do you guys think?

**Comment by chuwy** [comment]
Hey @BenFradet. This is indeed a good idea. 
We implied it as a part of https://github.com/snowplow/snowplow/issues/2824#issuecomment-235549134. I have a sbt template (for [sbt new](http://www.scala-sbt.org/0.13/docs/sbt-new-and-Templates.html)) which I wanted to propose as a basis for all newly created and existing snowplow projects. It includes a `build.sbt` and project files based on current our current projects. The only project with `build.sbt` is [sauna](https://github.com/snowplow/sauna), I think you may take it as an example.

**Comment by chuwy** [comment]
@alexanderdean can you create a project in snowplow organization for sbt template I mentioned above (it is incoplete though)?

**Comment by alexanderdean** [comment]
Sure thing @chuwy, what do you want the repo called?

**Comment by BenFradet** [comment]
@chuwy ok thanks, I'll follow the guidelines used in sauna

**Comment by chuwy** [comment]
@alexanderdean currently it called `snowplow-scala-skeleton.g8` (g8 is by [convention](https://github.com/foundweekends/giter8/wiki/giter8-templates)). We can come up with better name taking in account that we probably may want to have two these templates:

* for libs, like SCE, iglu-client etc
* for apps, like schema-guru, igluctl

These two types of apps usually have very different configuration, especially for deployment: release manager, sbt assembly, maven publishing etc.

**Comment by alexanderdean** [comment]
How about:

* `snowplow-scala-app-skeleton.g8`
* `snowplow-scala-lib-skeleton.g8`

although is skeleton implied by .g8 - what else is a .g8 but a skeleton?

**Comment by chuwy** [comment]
Agree. It is always skeleton/template. Let's call it `snowplow-scala-app.g8` and `snowplow-scala-lib.g8`.

**Comment by alexanderdean** [comment]
Done:

* https://github.com/snowplow/snowplow-scala-app.g8
* https://github.com/snowplow/snowplow-scala-lib.g8

You have admin access @chuwy.

**Comment by BenFradet** [comment]
Done with the port, reviews welcome!

**Comment by alexanderdean** [comment]
Cool! Will review. In the meantime, please crack on with the EmrEtlRunner and StorageLoader changes...

**Comment by BenFradet** [comment]
yup

**Comment by BenFradet** [comment]
I'm wondering if I shouldn't ditch typesafe for simple application arguments given that there are only 4 of them (input, good output, bad output, iglu resolver), what do you think?

File: 4-storage/relational-database-shredder/src/main/scala/com.snowplowanalytics.snowplow.storage/spark/ShredJob.scala [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+   * @param outFolder shredded/good/run=xxx
+   * @return The shredded types output path
+   */
+  def getShreddedTypesOutputPath(outFolder: String): String = {
+    val shreddedTypesSubdirectory = "shredded-types"
+    s"${outFolder}${if (outFolder.endsWith("/")) "" else "/"}${shreddedTypesSubdirectory}"
+  }
+}
+
+/** Case class representing shredded events. */
```
Comment (alexanderdean) [inline 4-storage/relational-database-shredder/src/main/scala/com.snowplowanalytics.snowplow.storage/spark/ShredJob.scala]: Can we use this opportunity to add @tparams for all of the case class properties

File: 4-storage/relational-database-shredder/src/main/scala/com.snowplowanalytics.snowplow.storage/spark/ShredJob.scala [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+case class Shredded(
+  eventId: String,
+  eventFingerprint: String,
+  shreds: List[JsonSchemaPair],
+  originalLine: String
+)
+
+/** Case class represeting good events and synthetic duplicates. */
```
Comment (alexanderdean) [inline 4-storage/relational-database-shredder/src/main/scala/com.snowplowanalytics.snowplow.storage/spark/ShredJob.scala]: Can we use this opportunity to add @tparams for all of the case class properties

File: 4-storage/relational-database-shredder/src/main/scala/com.snowplowanalytics.snowplow.storage/spark/ShredJob.scala [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+case class Event(eventId: String, newEventId: Option[String], shredded: Shredded)
+
+/** The Snowplow Shred job, written in Spark. */
```
Comment (alexanderdean) [inline 4-storage/relational-database-shredder/src/main/scala/com.snowplowanalytics.snowplow.storage/spark/ShredJob.scala]: Can we use this opportunity to add @tparams for all of the case class properties

File: 4-storage/relational-database-shredder/src/main/scala/com.snowplowanalytics.snowplow.storage/spark/ShredJob.scala [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+  import spark.implicits._
+  @transient private val sc: SparkContext = spark.sparkContext
+
+  // Job configuration
+  private val shredConfig = ShredJobConfig.loadConfigFrom(config).fold(
+    e => throw FatalEtlError(e.map(_.toString)),
+    c => c
+  )
+
+  def run(): Unit = {
```
Comment (alexanderdean) [inline 4-storage/relational-database-shredder/src/main/scala/com.snowplowanalytics.snowplow.storage/spark/ShredJob.scala]: Can we get a scaladoc on this method

File: 4-storage/relational-database-shredder/src/main/scala/com.snowplowanalytics.snowplow.storage/spark/ShredJobConfig.scala [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+package com.snowplowanalytics.snowplow.storage.spark
+
+import com.fasterxml.jackson.databind.JsonNode
+import com.snowplowanalytics.iglu.client.Resolver
+import com.snowplowanalytics.iglu.client.validation.ProcessingMessageMethods._
+import com.snowplowanalytics.snowplow.storage.spark.utils.JsonUtils
+import com.typesafe.config.Config
+import scalaz._
+import Scalaz._
+
```
Comment (alexanderdean) [inline 4-storage/relational-database-shredder/src/main/scala/com.snowplowanalytics.snowplow.storage/spark/ShredJobConfig.scala]: See comment above about imports

File: 4-storage/relational-database-shredder/src/main/scala/com.snowplowanalytics.snowplow.storage/spark/ShredJobConfig.scala [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+            .toProcessingMessage
+            .toValidationNel: ValidatedNel[JsonNode]
+        reso <- Resolver.parse(json)
+      } yield reso
+    }
+
+    (inFolder.toValidationNel |@| outFolder.toValidationNel |@| badFolder.toValidationNel |@|
+      igluResolver) { ShredJobConfig(_, _, _, _) }
+  }
+}
```
Comment (alexanderdean) [inline 4-storage/relational-database-shredder/src/main/scala/com.snowplowanalytics.snowplow.storage/spark/ShredJobConfig.scala]: I think the use of typesafe Config is fine

File: 4-storage/relational-database-shredder/src/main/scala/com.snowplowanalytics.snowplow.storage/spark/inputs/EnrichedEventLoader.scala [hunk 8]
```diff
@@ -1,49 +1,40 @@
...
-package inputs
+package com.snowplowanalytics.snowplow.storage.spark.inputs
 
-// Java
 import java.util.UUID
 
-// Joda-Time
+import com.snowplowanalytics.snowplow.enrich.common.{ValidatedEnrichedEvent, ValidatedString}
+import com.snowplowanalytics.snowplow.enrich.common.outputs.EnrichedEvent
+import com.snowplowanalytics.util.Tap._
```
Comment (alexanderdean) [inline 4-storage/relational-database-shredder/src/main/scala/com.snowplowanalytics.snowplow.storage/spark/inputs/EnrichedEventLoader.scala]: Can we use this opportunity to get rid of Tap? It was a bad idea by me to introduce it to Snowplow years ago and it's easily replaceable with an idiomatic Scala block.

File: 4-storage/relational-database-shredder/src/test/resources/log4j.properties [hunk 9]
```diff
@@ -0,0 +1,8 @@
+log4j.rootCategory=WARN, file
```
Comment (alexanderdean) [inline 4-storage/relational-database-shredder/src/test/resources/log4j.properties]: Is this file required by Spark for testing? Can we add a comment line to it explaining the rationale for adding?

File: 4-storage/relational-database-shredder/src/test/scala/com.snowplowanalytics.snowplow.storage.spark/BeforeAfterAll.scala [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+ * See the Apache License Version 2.0 for the specific language governing permissions and
+ * limitations there under.
+ */
+package com.snowplowanalytics.snowplow.storage.spark
+
+import org.specs2.mutable.SpecificationLike
+import org.specs2.specification.Fragments
+import org.specs2.specification.Step
+
+trait BeforeAfterAll extends SpecificationLike {
```
Comment (alexanderdean) [inline 4-storage/relational-database-shredder/src/test/scala/com.snowplowanalytics.snowplow.storage.spark/BeforeAfterAll.scala]: `BeforeAfterAll` sounds rather cryptic. Can we at least add a scaladoc for it?

File: 4-storage/relational-database-shredder/src/test/scala/com.snowplowanalytics.snowplow.storage.spark/ShredJobSpec.scala [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+    val job = ShredJob(spark, config)
+    job.run()
+    deleteRecursively(input)
+  }
+
+  override def afterAll(): Unit = {
+    super.afterAll()
+    dirs.deleteAll()
+  }
+}
```
Comment (alexanderdean) [inline 4-storage/relational-database-shredder/src/test/scala/com.snowplowanalytics.snowplow.storage.spark/ShredJobSpec.scala]: LGTM

File: 4-storage/relational-database-shredder/src/test/scala/com.snowplowanalytics.snowplow.storage.spark/SparkSpec.scala [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+
+/**
+ * Trait to mix in in every spec which has to run a Spark job.
+ * Create a spark session before the spec and delete it afterwards.
+ */
+trait SparkSpec extends BeforeAfterAll {
+  def appName: String
+
+  private val conf = new SparkConf()
+    .setMaster("local[1]")
```
Comment (alexanderdean) [inline 4-storage/relational-database-shredder/src/test/scala/com.snowplowanalytics.snowplow.storage.spark/SparkSpec.scala]: Can we add a comment just reminding what local[1] signifies?

File: 4-storage/relational-database-shredder/src/main/scala/com.snowplowanalytics.snowplow.storage/spark/ShredJob.scala [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
+      .flatMap { case (k, vs) => vs.take(1) }
+      .cache()
+
+    // Count synthetic duplicates, defined as events with the same id but different fingerprints
+    val syntheticDupes = good
+      .groupBy(_.eventId)
+      .filter { case (k, vs) => vs.size > 1 }
+      .map { case (k, vs) => (k, vs.size.toLong) }
+
+    // Join the properly-formed events with the synthetic duplicates, generate a new event id for
```
Comment (alexanderdean) [inline 4-storage/relational-database-shredder/src/main/scala/com.snowplowanalytics.snowplow.storage/spark/ShredJob.scala]: "event ID" not "event id"

File: 4-storage/relational-database-shredder/src/main/scala/com.snowplowanalytics.snowplow.storage/spark/ShredJob.scala [hunk 14]
```diff
@@ -0,0 +1,323 @@
...
+        getDuplicationContextHierarchy(e.eventId, e.newEventId, e.shredded.originalLine)
+      }
+
+    // Ready the events for database load
+    val events = goodWithSyntheticDupes
+      .map(e => alterEnrichedEvent(e.shredded.originalLine, e.newEventId))
+    if (!events.isEmpty)
+      events.saveAsTextFile(getAlteredEnrichedOutputPath(shredConfig.outFolder))
+
+    // Update the shredded jsons with the new deduplicated event IDs and stringify
```
Comment (alexanderdean) [inline 4-storage/relational-database-shredder/src/main/scala/com.snowplowanalytics.snowplow.storage/spark/ShredJob.scala]: Let's capitalize JSON/JSONs throughout this file

File: 4-storage/relational-database-shredder/project/BuildSettings.scala [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+import sbt._
+import Keys._
+
+object BuildSettings {
+
+  lazy val buildSettings = Seq[Setting[_]](
+    organization  := "com.snowplowanalytics",
+    scalaVersion  := "2.10.4",
+    scalacOptions := compilerOptions,
+    parallelExecution in Test := false, // Parallel tests cause havoc with Spark
```
Comment (BenFradet) [inline 4-storage/relational-database-shredder/project/BuildSettings.scala]: Yup, I've had issues with this in the past as in [spark-kafka-writer](https://github.com/benfradet/spark-kafka-writer) for example.
Though, tbh, I haven't tried with those particular tests.

File: 4-storage/relational-database-shredder/src/main/scala/com.snowplowanalytics.snowplow.storage/spark/ShredJob.scala [hunk 16]
```diff
@@ -0,0 +1,323 @@
...
+import com.snowplowanalytics.iglu.client.{JsonSchemaPair, ProcessingMessageNel, Resolver, SchemaKey}
+import com.snowplowanalytics.iglu.client.validation.ProcessingMessageMethods._
+import com.snowplowanalytics.snowplow.enrich.common.FatalEtlError
+import com.snowplowanalytics.snowplow.enrich.common.utils.shredder.Shredder
+import com.snowplowanalytics.snowplow.enrich.common.outputs.{BadRow, EnrichedEvent}
+import com.typesafe.config.{Config, ConfigFactory}
+import org.apache.spark.{SparkConf, SparkContext}
+import org.apache.spark.serializer.KryoSerializer
+import org.apache.spark.sql.{SQLContext, SparkSession}
+import scalaz._
```
Comment (BenFradet) [inline 4-storage/relational-database-shredder/src/main/scala/com.snowplowanalytics.snowplow.storage/spark/ShredJob.scala]: I don't have a strong opinion for either.
This approach just seems more compact and gets rid of the import clutter faster imho.

File: 4-storage/relational-database-shredder/src/test/resources/log4j.properties (hunk 9 above)
Comment (BenFradet) [inline 4-storage/relational-database-shredder/src/test/resources/log4j.properties]: It isn't required, without it everything is logged to stdout and clutters the test output however.

**Comment by BenFradet** [comment]
Thanks for your review!

File: 4-storage/relational-database-shredder/src/main/scala/com.snowplowanalytics.snowplow.storage/spark/ShredJobConfig.scala (hunk 7 above)
Comment (BenFradet) [inline 4-storage/relational-database-shredder/src/main/scala/com.snowplowanalytics.snowplow.storage/spark/ShredJobConfig.scala]: I was afraid it might be a bit too much since we only need 4 config values.
I was thinking we might pass them as command line arguments as was the case with the scalding job, what do you think?

File: 4-storage/relational-database-shredder/project/BuildSettings.scala (hunk 15 above)
Comment (alexanderdean) [inline 4-storage/relational-database-shredder/project/BuildSettings.scala]: Cool makes sense

File: 4-storage/relational-database-shredder/src/main/scala/com.snowplowanalytics.snowplow.storage/spark/ShredJobConfig.scala (hunk 7 above)
Comment (alexanderdean) [inline 4-storage/relational-database-shredder/src/main/scala/com.snowplowanalytics.snowplow.storage/spark/ShredJobConfig.scala]: Oh, I see what you mean. Yes, we should use command line arguments as in the Scalding job...

**Comment by BenFradet** [comment]
updated :+1:

**Comment by alexanderdean** [comment]
Great thanks @BenFradet - ready for @chuwy to do the next review?

**Comment by BenFradet** [comment]
yup

**Comment by BenFradet** [comment]
Downgrading to 1.6.1

**Comment by BenFradet** [comment]
done :+1:

**Comment by alexanderdean** [comment]
@chuwy - could you prioritize doing this review for @BenFradet ?

**Comment by chuwy** [comment]
Whoops. Missed it. Will review ASAP.

**Comment by BenFradet** [comment]
This isn't a real priority since I'm gonna need to revert back to Spark 2.0.

**Comment by BenFradet** [comment]
@chuwy rebumped to spark 2.0.2 :+1:

File: .travis/deploy_relational_database_shredder.sh [hunk 17]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
+
+tag=$1
+
+project_name="relational_database_shredder/"
+project_name_len=${#project_name}
```
Comment (BenFradet) [inline .travis/deploy_relational_database_shredder.sh]: @chuwy curious as to what you think about that, that would remove the length of the beginning of the tag being hardcoded?

Comment (chuwy) [inline .travis/deploy_relational_database_shredder.sh]: I like it! Hard-coded length is confusing.

Comment (BenFradet) [inline .travis/deploy_relational_database_shredder.sh]: Do you think it's worth it to modify the other scripts in the same fashion?

**Comment by BenFradet** [comment]
The move to 2.11 depends on #3062

**Comment by BenFradet** [comment]
Need to remove [ValidatedNel](https://github.com/snowplow/snowplow/pull/3037/files#diff-265ca3cab7f2e4d6d94bbc1e4ea33360L46) in favor of [ValidatedNelMessage](https://github.com/snowplow/snowplow/blob/master/3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/package.scala#L150)

**Comment by BenFradet** [comment]
Closing in favor of #3094

**Comment by alexanderdean** [comment]
De-scheduling

**Comment by dashirov-ga** [comment]
Was this de-scheduled to synchronize breaking changes with existing EMR/Hadoop tooling or for some other reason? Was this discussed anywhere on discourse or elsewhere?

**Comment by alexanderdean** [comment]
It was de-scheduled because it was closed, being superceded by #3094 per @BenFradet's comment above.
